src/mlpack/core/tree/binary_space_tree/rp_tree_max_split.hpp

```diff
@@ -0,0 +1,109 @@
+/**
+ * @file rp_tree_max_split.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Definition of class (RPTreeMaxSplit) to split a binary space partition
+ * tree.
+ */
+#ifndef MLPACK_CORE_TREE_BINARY_SPACE_TREE_RP_TREE_MAX_SPLIT_HPP
+#define MLPACK_CORE_TREE_BINARY_SPACE_TREE_RP_TREE_MAX_SPLIT_HPP
+
+#include <mlpack/core.hpp>
+
+namespace mlpack {
+namespace tree /** Trees and tree-building procedures. */ {
+
+/**
+ * This class splits a node by a random hyperplane. In order to choose the
+ * hyperplane we need to choose the normal to the hyperplane and the position
+ * of the hyperplane i.e. the scalar product of the normal and a point.
+ *
+ * A point will be assigned to the left subtree if the product of
+ * this point and the normal is less or equal to the split value (i.e. the
+ * position of the hyperplane).
+ */
+template<typename BoundType, typename MatType = arma::mat>
+class RPTreeMaxSplit
+{
+ public:
+  typedef typename MatType::elem_type ElemType;
+  struct SplitInfo
+  {
+    //! The normal vector to the hyperplane that splits the node.
+    arma::Col<ElemType> direction;
+    //! The value according to which the node is being split.
+    ElemType splitVal;
+  };
+  /**
+   * Split the node by a random hyperplane.
+   *
+   * @param bound The bound used for this node.
+   * @param data The dataset used by the binary space tree.
+   * @param begin Index of the starting point in the dataset that belongs to
+   *    this node.
+   * @param count Number of points in this node.
+   * @param splitInfo An information about the split. This information contains
+   *    the direction and the value.
+   */
+  static bool SplitNode(const BoundType& /*bound*/,
+                        MatType& data,
+                        const size_t begin,
+                        const size_t count,
+                        SplitInfo& splitInfo);
+
+  /**
+   * Indicates that a point should be assigned to the left subtree.
+   *
+   * @param point The point that is being assigned.
+   * @param splitInfo An information about the split.
+   */
+  template<typename VecType>
+  static bool AssignToLeftNode(
+    const VecType& point,
+    const SplitInfo& splitInfo)
+  {
+    return (arma::dot(point, splitInfo.direction) <= splitInfo.splitVal);
```

Comment:
I spent a while thinking about this.  The inner product is defined for L2 space but only for L2 space---in other spaces like L1, L3, L_inf, etc., it is not defined.  I think this is why the paper restricts itself to considering L2.  But I don't think that using the dot product like this in other spaces is a problem.  Let me know if you think otherwise.